src/cbapi/psc/livequery/query.py

```diff
@@ -291,6 +291,148 @@ def submit(self):
+            self._query_builder = q
+        else:
+            self._query_builder.where(q, **kwargs)
+        return self
+
+    def and_(self, q=None, **kwargs):
+        """Add a conjunctive filter to this query.
+
+        :param q: Query string or `solrq.Q` object
+        :param kwargs: Arguments to construct a `solrq.Q` with
+        :return: Query object
+        :rtype: :py:class:`Query`
+        """
+        if not q and not kwargs:
+            raise ApiError(".and_() expects a string, a solrq.Q, or kwargs")
+
+        self._query_builder.and_(q, **kwargs)
+        return self
+
+    def or_(self, q=None, **kwargs):
+        """Add a disjunctive filter to this query.
+
+        :param q: `solrq.Q` object
+        :param kwargs: Arguments to construct a `solrq.Q` with
+        :return: Query object
+        :rtype: :py:class:`Query`
+        """
+        if not q and not kwargs:
+            raise ApiError(".or_() expects a solrq.Q or kwargs")
+
+        self._query_builder.or_(q, **kwargs)
+        return self
+
+    def not_(self, q=None, **kwargs):
+        """Adds a negated filter to this query.
+
+        :param q: `solrq.Q` object
+        :param kwargs: Arguments to construct a `solrq.Q` with
+        :return: Query object
+        :rtype: :py:class:`Query`
+        """
+
+        if not q and not kwargs:
+            raise ApiError(".not_() expects a solrq.Q, or kwargs")
+
+        self._query_builder.not_(q, **kwargs)
+        return self
+
+    def sort_by(self, key, direction="ASC"):
+        """Sets the sorting behavior on a query's results.
+
+        Example::
+
+        >>> cb.select(Result).run_id(my_run).where(username="foobar").sort_by("uid")
+
+        :param key: the key in the schema to sort by
+        :param direction: the sort order, either "ASC" or "DESC"
+        :rtype: :py:class:`ResultQuery`
+        """
+        self._sort.update({"field": key, "order": direction})
+        return self
+
+    def _build_request(self, start, rows):
+        request = {"start": start, "query": self._query_builder._collapse()}
+
+        if rows != 0:
+            request["rows"] = rows
+        if self._sort:
+            request["sort"] = [self._sort]
+
+        return request
+
+    def _count(self):
+        if self._count_valid:
+            return self._total_results
+
+        url = self._doc_class.urlobject_history.format(
+            self._cb.credentials.org_key
+        )
+        request = self._build_request(start=0, rows=0)
+        resp = self._cb.post_object(url, body=request)
+        result = resp.json()
+
+        self._total_results = result["num_found"]
+        self._count_valid = True
+
+        return self._total_results
+
+    def _perform_query(self, start=0, rows=0):
+        url = self._doc_class.urlobject_history.format(
+            self._cb.credentials.org_key
+        )
+        current = start
+        numrows = 0
+        still_querying = True
+        while still_querying:
```

Comment:
What if the query never finds results then this looks like an infinite loop

Reply:
See line 432. still_querying gets set to False once "current" reaches "_total_results." I'm pretty sure I copied this logic from someplace else.

Reply:
ah! I forget that if total_results returns 0 then current will also be 0 which will fulfill the >=. I over thought the condition